README.md
# Weekly PPT

It reads from ppt on a server, sync the data to databases, or generate a ppt from the databases. The structure of databases are defined in eztools.

ChangeLog
1.0	20190930	Allen Tse
	Uploaded to github.com and open-sourced.
2.2	20191030	Allen Tse
	Bundled import and export. Export only for managers.


acc_test.go
package main

import (
	"fmt"
	"testing"

	"github.com/bon-ami/eztools"
)

func TestGetUsr(t *testing.T) {
	logger, db := prepareEnv()
	if logger != nil {
		defer logger.Close()
	}
	var (
		idE, idF   int
		errE, errF error
		output     string
	)

	//sub test env
	if db != nil {
		defer db.Close()
		idE, errE = getUsrFromEnv(db)
	}

	//sub test cfg file
	idF, output, errF = parseUsrFromFile(func(buf,
		id string) (output string, breaking bool) {
		return "matched", true
	}, func(buf, id string) (output string, breaking bool) {
		return
	}, 0)
	if errE != nil || errF != nil {
		fmt.Println(errE)
		fmt.Println(errF)
		t.Fail()
	}
	if db != nil {
		dispRes(t, "from env.", idE, output)
	}
	dispRes(t, "from cfg.", idF, output)
}

func dispRes(t *testing.T, from string, id int, output string) {

	switch id {
	case 0:
		t.Error(from + " NO ID")
	case eztools.InvalidID:
		t.Error(from + " Invalid ID")
	default:
		fmt.Println(from+" ID ", id, " ", output)
	}
}


mng_test.go
package main

import (
	"database/sql"
	"testing"
)

func TestMng(t *testing.T) {
	mainLoop(func(*sql.DB, int) (su, mng bool) {
		return false, true
	})
}


inex_test.go
package main

import (
	"fmt"
	"io"
	"os"
	"testing"
	"time"
)

func TestInEx(t *testing.T) {
	logger, db := prepareEnv()
	if logger != nil {
		defer logger.Close()
	}

	const (
		fn = "\\\\10.75.10.81\\Share\\Team\\FW&APP\\Documents\\test.ppt"
		sl = 30 * time.Second
	)
	var p2d ppt2db
	p2d.init(db, week, nil)
	fmt.Println("")
	fp, err := os.OpenFile(fn, os.O_RDWR, 0666)
	if err != nil {
		fmt.Println(fn + " NOT FOUND!")
		t.FailNow()
	}
	fi, err := os.Stat(fn)
	if err != nil {
		fmt.Println(fn + " NOT readable!")
		t.FailNow()
	}
	if fp != nil {
		rdByRD(db, fp, fi.Size(), &p2d)
		fmt.Println("")
		fmt.Println(fn + " locked for " + sl.String())
		time.Sleep(sl)
		_, err = fp.Seek(0, io.SeekStart)
		if err == nil {
			err = fp.Truncate(0)
			if err == nil {
				wrByWR(db, fp)
			}
		}
		fp.Close()
		if err != nil {
			t.Fail()
		}
	}
}


acc.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"errors"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/bon-ami/eztools"
	"github.com/bon-ami/eztools/contacts"
)

const (
	cfgFieldUsr  = "user"
	cfgFileName  = "WeeklyPpt.cfg"
	cfgSepFields = ","
	cfgSepValues = "="
	cfgFileMode  = 0644
)

var (
	errValidityNotChecked = errors.New("Not able to check validity")
	errNotDefined         = errors.New("Not defined or configured")
	errNotSaved           = errors.New("Result not saved")
	cfgEnvUsrs            = [][]string{
		{"windows", "USERNAME"},
		{"linux", "USER"}}
)

func validateUsrID(db *sql.DB, id int) (user string, err error) {
	searched, err := eztools.Search(db, eztools.TblCONTACTS,
		eztools.FldID+"="+strconv.Itoa(id),
		[]string{eztools.FldNAME}, "")
	if err == nil {
		return searched[0][0], nil
	}
	return
}

func getCfgFileName() string {
	home, err := os.UserHomeDir()
	if err != nil {
		eztools.LogErr(err)
		return cfgFileName
	}
	return filepath.Join(home, cfgFileName)
}

//func getCfgFile(mode int) (*os.File, error) {
//return os.OpenFile(getCfgFileName(), mode, 0664)
//}

func getUsrFromSys() (name string) {
	for _, v := range cfgEnvUsrs {
		if runtime.GOOS == v[0] {
			return os.Getenv(v[1])
		}
	}
	return
}

func getUsrFromEnv(db *sql.DB) (id int, err error) {
	err = eztools.ErrNoValidResults
	for _, v := range cfgEnvUsrs {
		if runtime.GOOS == v[0] {
			id, err = contacts.GetIDFromAllNames(db,
				os.Getenv(v[1]))
			if err == nil && id == eztools.InvalidID {
				err = eztools.ErrNoValidResults
			}
			if err != nil {
				eztools.LogErr(err)
			}
			break
		}
	}
	return
}

func createHash(key string) string {
	hasher := md5.New()
	_, err := hasher.Write([]byte(key))
	if err != nil {
		eztools.LogErr(err)
		return key
	}
	return hex.EncodeToString(hasher.Sum(nil))
}

func encrypt(data []byte, passphrase string) []byte {
	block, _ := aes.NewCipher([]byte(createHash(passphrase)))
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		eztools.LogErrPrint(err)
		return nil
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		eztools.LogErrPrint(err)
		return nil
	}
	ciphertext := gcm.Seal(nonce, nonce, data, nil)
	return ciphertext
}

func decrypt(data []byte, passphrase string) []byte {
	key := []byte(createHash(passphrase))
	block, err := aes.NewCipher(key)
	if err != nil {
		eztools.LogErrPrint(err)
		return nil
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		eztools.LogErrPrint(err)
		return nil
	}
	nonceSize := gcm.NonceSize()
	nonce, ciphertext := data[:nonceSize], data[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		eztools.LogErrPrint(err)
		return nil
	}
	return plaintext
}

type execCmpFieldUsr func(buf, input string) (output string, breaking bool)

func parseUsrFromFile(matched execCmpFieldUsr, unmatched execCmpFieldUsr,
	idIn int) (idOut int, output string, err error) {
	idOut = eztools.InvalidID
	buf, err := ioutil.ReadFile(getCfgFileName())
	if err != nil {
		eztools.LogErr(err)
		return
	}
	plaintext := decrypt(buf, getUsrFromSys())
	if plaintext == nil {
		err = eztools.ErrOutOfBound
		return
	}
	var (
		slc []string
	)
	/*scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		//read by line
		slc = strings.Split(scanner.Text(), "=")
	slc = strings.Split(buf, "=")*/
	var execBreak bool
	idInStr := strconv.Itoa(idIn)
	for _, fields := range strings.Split(string(plaintext), cfgSepFields) {
		slc = strings.Split(fields, cfgSepValues)
		if len(slc) == 2 && slc[0] == cfgFieldUsr {
			if len(slc[1]) < 1 {
				err = errNotDefined
			} else {
				idOut, err = strconv.Atoi(slc[1])
			}
			output, execBreak = matched(output, idInStr)
		} else {
			output, execBreak = unmatched(output, fields)
		}
		if execBreak {
			break

		}
	}
	return
}

func getUsrFromfile(db *sql.DB) (id int, fixed bool, err error) {
	/*file, err := getCfgFile(os.O_RDONLY)
	if err != nil {
		return
	}
	defer file.Close()*/
	id, err = getUsrFromEnv(db)
	if err == nil {
		fixed = true
		return
	}
	id, _, err = parseUsrFromFile(func(buf,
		id string) (output string, breaking bool) {
		return "", true
	}, func(buf, id string) (output string, breaking bool) {
		return
	}, 0)
	/* if err = scanner.Err(); err != nil {
		return
	}
	if len(usr) < 1 {
		return 0, fixed, errNotDefined
	}
	id, err = strconv.Atoi(usr)*/
	return
}

//func chgUsrFromfile( /*cfg string,*/ id int) (buf string, err error) {
/*file, err := getCfgFile(os.O_CREATE | os.O_RDWR)
if err != nil {
	return
}
defer file.Close()
var (
	slc  []string
	line string
)
scanner := bufio.NewScanner(file)
for scanner.Scan() {
	line = scanner.Text()
	slc = strings.Split(line, "=")
	if len(slc) == 2 && slc[0] == cfgFieldUsr {
		slc[1] = strconv.Itoa(id)
		buf = append(buf, strings.Join(slc, "="))
	} else {
		buf = append(buf, line)
	}
}*/
/*_, buf, err = parseUsrFromFile(func(buf, idIn string) (output string, breaking bool) {
	output = buf
	if len(buf) > 0 {
		output += cfgSepFields
	}
	return output + cfgFieldUsr + cfgSepValues + idIn, false
}, func(buf, field string) (output string, breaking bool) {
	return buf + field, false
}, id)*/
/*if err = scanner.Err(); err != nil {
	return
}*/
//return
//}

func putUsrTofile(id int) (err error) {
	_, buf, err := parseUsrFromFile(func(buf,
		idIn string) (output string, breaking bool) {
		output = buf
		if len(buf) > 0 {
			output += cfgSepFields
		}
		return output + cfgFieldUsr + cfgSepValues + idIn, false
	}, func(buf, field string) (output string, breaking bool) {
		return buf + field, false
	}, id)
	//buf, err := chgUsrFromfile(id)
	if err != nil && !os.IsNotExist(err) { //err != os.ErrNotExist {
		return
	}
	if len(buf) < 1 {
		buf = cfgFieldUsr + cfgSepValues + strconv.Itoa(id)
	}

	err = ioutil.WriteFile(getCfgFileName(), encrypt([]byte(buf),
		getUsrFromSys()), cfgFileMode)
	/*file, err := getCfgFile(os.O_CREATE | os.O_TRUNC | os.O_WRONLY)
	if err != nil {
		return
	}
	defer file.Close()
	for _, line := range buf {
		_, err = file.WriteString(line + "\n")
		if err != nil {
			break
		}
	}*/
	return
}

func chgUsr(db *sql.DB) (int, error) {
	selected, err := eztools.Search(db, eztools.TblCONTACTS, "",
		[]string{eztools.FldID, eztools.FldNAME, "ldap"}, "")
	if err != nil {
		return eztools.InvalidID, err
	}
	id := eztools.ChooseInts(selected, "Choose your ID")
	if id == eztools.InvalidID {
		return id, errNotDefined
	}
	pw := eztools.PromptPwd("Password")
	if len(pw) < 1 {
		return id, errValidityNotChecked
	}
	var ldap string
	for _, v := range selected {
		if v[0] == strconv.Itoa(id) {
			ldap = v[2]
			break
		}
	}
	err = eztools.Authenticate(db, ldap, pw)
	if err != nil {
		eztools.LogErrPrint(err)
		return id, errValidityNotChecked
	}
	err = putUsrTofile(id)
	if err != nil {
		return id, errNotSaved
	}
	return id, nil
}

func getUsr(db *sql.DB) (int, bool, error) {
	id, fixed, err := getUsrFromfile(db)
	if err == nil {
		usr, err := validateUsrID(db, id)
		if err == nil {
			eztools.ShowStr("Welcome, " + usr + ". ")
			return id, fixed, nil
		}
		return id, fixed, errValidityNotChecked
	}
	eztools.ShowStrln("No valid account set.")
	return id, fixed, err
}


chk.go
package main

import (
	"database/sql"
	"strconv"

	"github.com/bon-ami/eztools"
)

//parameters: member[0]:ID; [1]:name
func listTask1Member1Week(db *sql.DB, member []string, week int, weekDesc, table string, add bool) (skipMember bool) {
	for {
		selected, err := eztools.Search(db, eztools.TblWEEKLYTASKWORK,
			"contact="+member[0]+" AND week="+strconv.Itoa(week),
			[]string{eztools.FldID, eztools.FldSTR, "section"}, " ORDER BY section")
		if err != nil {
			eztools.LogErrPrint(err)
			return false
		}
		if len(selected) < 1 {
			eztools.ShowStrln("No current tasks for " + member[1] +
				", " + weekDesc + ".")
			break
		} else {
			eztools.ShowStrln("Current tasks for " + member[1] +
				", " + weekDesc + " listed as below:")
			for i, work := range selected {
				if len(work[1]) < 1 || len(work[2]) < 1 {
					eztools.LogPrint("zero length for " +
						member[0] + "'s task of week " +
						strconv.Itoa(week))
					continue
				}
				secID, err := eztools.GetPairStr(db, table, work[2])
				if err != nil {
					eztools.LogErrWtInfo("Processing member "+member[0]+"'s task in plan", err)
					continue
				}
				secDesc, err := eztools.GetPairStr(db, eztools.TblWEEKLYTASKTITLES, secID)
				if err != nil {
					eztools.LogErrWtInfo("Processing member "+member[0]+"'s task in titles", err)
					continue
				}
				des, err := eztools.GetPairStr(db, eztools.TblWEEKLYTASKDESC, work[1])
				if err != nil {
					eztools.LogErrWtInfo("Processing member "+member[0]+"'s description", err)
					continue
				}
				eztools.ShowStrln("\t" + strconv.Itoa(i) + ": " + secDesc + ". " + des)
			}
			if add {
				var skipWeek bool
				skipWeek, skipMember = queryTasks(db, table, member, selected)
				if skipMember {
					return
				}
				if skipWeek {
					break
				}
			}
		}
		if !add {
			return
		}
	}
	if !add {
		return
	}
	eztools.ShowStrln("Now, to add tasks.")
	for {
		horizontal(20)
		switch addTaskUI(db, member, week, weekDesc, table) {
		case eztools.ErrInvalidInput:
			//skipMember = true
			return
		case nil:
		default:
			return
		}
	}
}

func member4weeks2Lst(db *sql.DB, member []string) {
	listTask1Member1Week(db, member, week, "this week", eztools.TblWEEKLYTASKCURR, false)
	listTask1Member1Week(db, member, week+1, "next week", eztools.TblWEEKLYTASKNEXT, false)
}

func chk(db *sql.DB, acc int) {
	members4weeks(db, acc, member4weeks2Lst)
}


fll.go
package main

import (
	"database/sql"
	"strconv"

	"github.com/bon-ami/eztools"
)

func modTaskExec(db *sql.DB, id string, str int) {
	if err := eztools.UpdatePairWtParams(db,
		eztools.TblWEEKLYTASKWORK, id, strconv.Itoa(str)); err != nil {
		eztools.LogErr(err)
	}
}

//parameters: member[0]:ID; [1]:name;	id[0]: ID; [1]: str; [2]: section
func modTask(db *sql.DB, table string, member, id []string) {
	secStr := id[2]
	section, err := chooseSection(db, table,
		"Choose a section if you want to move this task. Invalid to keep current")
	if err == nil {
		prevSec, err := strconv.Atoi(id[2])
		if err == nil && prevSec != section {
			err = eztools.UpdateWtParams(db, eztools.TblWEEKLYTASKWORK,
				eztools.FldID+"="+id[0], []string{"section"},
				[]string{strconv.Itoa(section)}, false)
			if err != nil {
				eztools.LogErrPrint(err)
			} else {
				eztools.ShowStrln("moved.")
			}
			secStr = strconv.Itoa(section)
		}
	}
	descID, err := chooseOrAddTaskDesc(db, secStr)
	if err == nil {
		modTaskExec(db, id[0], descID)
	}
}

func chooseOrAddTaskDesc(db *sql.DB, section string) (descID int, err error) {
	works, err := eztools.Search(db, eztools.TblWEEKLYTASKDESC, "",
		[]string{"DISTINCT d.id", "d.str"},
		" d JOIN "+eztools.TblWEEKLYTASKWORK+
			" w ON w.section="+section+
			" AND w.str=d.id")
	if err != nil {
		eztools.LogErrPrint(err)
	} else {
		descID = eztools.ChooseInts(works,
			"Choose from tasks in existence (invalid to create one)")
		if descID != eztools.InvalidID {
			return
		}
	}
	descID, err = promptTaskDesc(db)
	return
}

func addTaskDesc(db *sql.DB, desc string) (descID int, err error) {
	descID, err = eztools.Locate(db, eztools.TblWEEKLYTASKDESC, desc)
	if err != nil {
		eztools.LogErrPrint(err)
		return
	}
	if descID != eztools.DefID {
		//already in existence
		return
	}
	descID, err = eztools.AddPairNoID(db, eztools.TblWEEKLYTASKDESC, desc)
	if err != nil {
		eztools.LogErrPrint(err)
		return
	}
	return
}

func promptTaskDesc(db *sql.DB) (descID int, err error) {
	desc := eztools.PromptStr("Input description for this task (invalid value to return)")
	if len(desc) < 1 {
		err = eztools.ErrInvalidInput
		return
	}
	return addTaskDesc(db, desc)
}

func chooseSection(db *sql.DB, table, desc string) (section int, err error) {
	sections, err := eztools.Search(db,
		eztools.TblWEEKLYTASKTITLES+","+table,
		eztools.TblWEEKLYTASKTITLES+".id="+table+".str",
		[]string{"DISTINCT " + eztools.TblWEEKLYTASKTITLES + ".id",
			eztools.TblWEEKLYTASKTITLES + ".str"}, "")
	if err != nil {
		eztools.LogErrPrint(err)
		return
	}
	section = eztools.ChooseInts(sections, desc)
	if section == eztools.InvalidID {
		err = eztools.ErrInvalidInput
		return
	}
	section, err = eztools.GetPairID(db, table, strconv.Itoa(section))
	if err != nil {
		eztools.LogErrPrint(err)
	}
	return
}

//parameters: member[0]:ID; [1]:name
func addTaskUI(db *sql.DB, member []string, wk int, weekDesc, table string) (err error) {
	section, err := chooseSection(db, table,
		"Choose one section for the new task for "+
			member[1]+" of "+weekDesc)
	if err != nil {
		return
	}
	secStr := strconv.Itoa(section)
	desc, err := chooseOrAddTaskDesc(db, secStr)
	if err != nil {
		eztools.LogErrPrint(err)
		return
	}
	err = addTask(db, strconv.Itoa(desc), member[0], secStr, strconv.Itoa(wk))
	if err != nil {
		eztools.LogErrPrint(err)
	}
	return
}

func promptID(notif string, len int) (id int, err error) {
	id, err = eztools.PromptInt("Which task to " + notif + "?")
	if err != nil || id < 0 || id >= len {
		return id, eztools.ErrInvalidInput
	}
	return
}

//parameters: member[0]:ID; [1]:name;	id[][0]: ID; [1]: str; [2]: section
func queryTasks(db *sql.DB, table string, member []string, id [][]string) (skipWeek, skipMember bool) {
	choices := []string{
		"Change a task above",                  //0
		"Delete a task above",                  //1
		"Skip modifying tasks above (default)", //2
		"Skip this member"}                     //3
	var (
		err error
		i   int
	)
	switch eztools.ChooseStrings(choices) {
	case 0:
		i, err = promptID("modify", len(id))
		if err == nil {
			modTask(db, table, member, id[i])
		}
	case 1:
		i, err = promptID("delete", len(id))
		if err == nil {
			err = eztools.DeleteWtID(db, eztools.TblWEEKLYTASKWORK, id[i][0])
		}
	case 3:
		skipMember = true
	default:
		skipWeek = true
	}
	if err != nil {
		eztools.LogErrPrint(err)
	}
	return
}

func member4weeks2Add(db *sql.DB, member []string) {
	//if week < 2 {
	//eztools.ShowStrln("No reference plan for the first week of a year.")
	//}
	if listTask1Member1Week(db, member, week, "this week", eztools.TblWEEKLYTASKCURR, true) {
		return
	}
	horizontal(10)
	listTask1Member1Week(db, member, week+1, "next week", eztools.TblWEEKLYTASKNEXT, true)
}

func fll(db *sql.DB, acc int) {
	members4weeks(db, acc, member4weeks2Add)
}


mnt.go
package main

import (
	"database/sql"
	"strconv"

	"github.com/bon-ami/eztools"
)

func listTitles1(db *sql.DB, tbl string) {
	pi, err := eztools.GetSortedPairsInt(db, tbl)
	if err != nil {
		eztools.LogErr(err)
		return
	}
	var (
		id, val int
		desc    string
	)
	for {
		if id, val, err = pi.Next(); err != nil {
			break
		}
		desc, err = eztools.GetPairStrFromInt(db, eztools.TblWEEKLYTASKTITLES, val)
		if err != nil {
			eztools.LogErr(err)
			continue
		}
		eztools.ShowStrln("ID " + strconv.Itoa(id) + ". Page " + strconv.Itoa(id/10) + ": " + desc)
	}
}

func chkTitle41Page(db *sql.DB, tbl, desc string) (int, error) {
	idC, err := eztools.PromptInt("ID for " + desc + " page")
	if err != nil {
		eztools.LogErr(err)
		return eztools.InvalidID, err
	}
	id, err := eztools.GetPairStrFromInt(db, tbl, idC)
	if err != nil {
		if err == eztools.ErrNoValidResults {
			err = nil
		} else {
			eztools.LogErr(err)
		}
	} else {
		eztools.ShowStrln("ID " + id + " already exists!")
		err = eztools.ErrInvalidInput
	}
	return idC, err
}

func addTitle(db *sql.DB) {
	var (
		idS, idI int
		err      error
	)
	str := eztools.PromptStr("Title name")
	if len(str) < 1 {
		return
	}
	idS, err = eztools.AddPairNoID(db, eztools.TblWEEKLYTASKTITLES, str)
	if err != nil {
		return
	}
	idI = eztools.InvalidID
	defer func() {
		if err != nil {
			//TODO: lock the table? Tx does not work, since idS needed for these two additions
			if idS != eztools.InvalidID {
				if err = eztools.DeleteWtID(db, eztools.TblWEEKLYTASKTITLES, strconv.Itoa(idS)); err != nil {
					eztools.LogErr(err)
				}
			}
			if idI != eztools.InvalidID {
				if err = eztools.DeleteWtID(db, eztools.TblWEEKLYTASKCURR, strconv.Itoa(idI)); err != nil {
					eztools.LogErr(err)
				}
			}
		}
	}()
	if idI, err = addTitle2Layout(db, eztools.TblWEEKLYTASKCURR, "current", idS); err != nil {
		return
	}
	_, err = addTitle2Layout(db, eztools.TblWEEKLYTASKNEXT, "next", idS)
}

func addTitle2Layout(db *sql.DB, table, desc string, title int) (idI int, err error) {
	idC, err := chkTitle41Page(db, table, desc)
	if err != nil {
		return
	}
	idI, err = eztools.AddPair(db, table, idC, strconv.Itoa(title))
	return
}

func chgTitle(db *sql.DB) {
	id, err := eztools.ChoosePair(db, eztools.TblWEEKLYTASKTITLES)
	if err != nil {
		eztools.LogErrPrint(err)
		return
	}
	str := eztools.PromptStr("Change it to?")
	if len(str) > 0 {
		err = eztools.UpdatePairWtParams(db,
			eztools.TblWEEKLYTASKTITLES, strconv.Itoa(id), str)
		if err != nil {
			eztools.LogErrPrint(err)
		}
	}
}

func chgLayout1Add(db *sql.DB, table string, title int) {
	if title == eztools.InvalidID {
		addTitle(db)
	} else {
		if _, err := addTitle2Layout(db, table, "", title); err != nil {
			eztools.LogErr(err)
		}
	}
}

func chgLayout1Mod(db *sql.DB, table string, id, newID int) {
	_, err := eztools.GetPairStrFromInt(db, table, newID)
	if err == nil {
		eztools.ShowStrln("ID already exists!")
		return
	}
	if err != eztools.ErrNoValidResults {
		eztools.LogErrPrint(err)
		return
	}
	idSOld := strconv.Itoa(id)
	idSNew := strconv.Itoa(newID)
	if err = eztools.UpdatePairID(db, table,
		idSOld, idSNew); err != nil {
		eztools.LogErrPrint(err)
		return
	}
	if err = eztools.UpdateWtParams(db, table, eztools.FldSECTION+"="+idSOld,
		[]string{eztools.FldSECTION},
		[]string{idSNew}, true); err != nil {
		eztools.LogErrPrint(err)
		return
	}
}

//operations are same on both curr and next tables
func chgLayout1(db *sql.DB, table string) {
	eztools.ShowStrln("Which title to change? (IDs are before colons)")
	id, err := eztools.ChoosePairOrAdd(db, table, true)
	if err != nil {
		eztools.LogErrPrint(err)
		return
	}
	if id == eztools.InvalidID {
		//add an item
		eztools.ShowStrln("Choose a title in existence or add one.")
		title, err := eztools.ChoosePairOrAdd(db, eztools.TblWEEKLYTASKTITLES, true)
		if err != nil {
			eztools.LogErrPrint(err)
			return
		}
		chgLayout1Add(db, eztools.TblWEEKLYTASKCURR, title)
		chgLayout1Add(db, eztools.TblWEEKLYTASKNEXT, title)
	} else {
		//move an item
		newID, err := eztools.PromptInt("new ID=")
		if err != nil {
			return
		}
		chgLayout1Mod(db, eztools.TblWEEKLYTASKCURR, id, newID)
		chgLayout1Mod(db, eztools.TblWEEKLYTASKNEXT, id, newID)
	}
}

func chgLayout(db *sql.DB) {
	//switch eztools.PromptStr("Current page or Next?") {
	//case "c", "C":
	chgLayout1(db, eztools.TblWEEKLYTASKCURR)
	//case "n", "N":
	//chgLayout1(db, eztools.TblWEEKLYTASKNEXT)
	//}
}

func mnt(db *sql.DB) {
	eztools.ShowStrln("Current week's pages,")
	listTitles1(db, eztools.TblWEEKLYTASKCURR)
	eztools.ShowStrln("Next week's pages,")
	listTitles1(db, eztools.TblWEEKLYTASKNEXT)
	choices := []string{"Return", //0
		"Add a title",    //1
		"Modify a title", //2
		"Delete a title", //3
		"Modify layout"}  //4
	c := eztools.ChooseStrings(choices)
	switch c {
	case 0:
	case 1:
		addTitle(db)
	case 2:
		chgTitle(db)
	case 4:
		chgLayout(db)
	default:
		eztools.ShowStrln("TODO")
	}
}


main.go
package main

import (
	"database/sql"
	"flag"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"time"

	_ "github.com/go-sql-driver/mysql"

	"github.com/bon-ami/eztools"
	"github.com/bon-ami/eztools/contacts"
)

var (
	ver  string
	week int
)

// return values: [][0] ID, [][1] Name
func getMembers(db *sql.DB, acc int) ([][]string, error) {
	if acc == eztools.InvalidID {
		acc, err := chgUsr(db)
		if (err != nil && err == errNotSaved) || acc == eztools.InvalidID {
			return nil, err
		}
	}
	return contacts.GetMembers(db, acc)

}

type member4weeksFn func(*sql.DB, []string)

func members4weeks(db *sql.DB, acc int, fc member4weeksFn) {
	members, err := getMembers(db, acc)
	if err != nil || members == nil {
		return
	}
	for _, member := range members {
		horizontal(30)
		fc(db, member)
	}
}

func addTaskInt(db *sql.DB, desc, member, sec, wk int) error {
	return addTask(db, strconv.Itoa(desc), strconv.Itoa(member), strconv.Itoa(sec), strconv.Itoa(wk))
}

func addTask(db *sql.DB, desc, member, sec, wk string) error {
	_, err := eztools.AddWtParamsUniq(db, eztools.TblWEEKLYTASKWORK,
		[]string{eztools.FldSTR, "contact", "section", "week"},
		[]string{desc, member, sec, wk}, false)
	return err
}

func horizontal(level int) {
	if level < 1 {
		level = 30
	}
	for i := 0; i < level; i++ {
		eztools.ShowStr("-")
	}
	eztools.ShowStrln("")
}

func chkManager(db *sql.DB, acc int) (su, mng bool) {
	switch acc {
	case 1: //super user
		return true, false
	case eztools.InvalidID:
		return false, false
	}
	searched, err := eztools.Search(db, eztools.TblTEAM,
		eztools.FldLEADER+"="+strconv.Itoa(acc), nil,
		" AND "+eztools.FldID+"<10")
	// IDs over 10 are for sub-teams
	if err != nil {
		eztools.LogErr(err)
		return false, false
	}
	if len(searched) > 0 {
		return false, true
	}
	return false, false
}

func prepareEnv() (logger *os.File, db *sql.DB) {
	_, week = time.Now().ISOWeek()
	eztools.ShowStrln("V" + ver + ". Now it is week " + strconv.Itoa(week))
	home, _ := os.UserHomeDir()
	logger, err := os.OpenFile(filepath.Join(home, "WeeklyPpt.log"), os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
	if err == nil {
		if err = eztools.InitLogger(logger); err != nil {
			eztools.ShowStrln(err.Error())
			logger.Close()
			logger = nil
		}
	} else {
		eztools.ShowStrln("Failed to open log file")
	}

	if len(ver) < 1 {
		ver = "dev"
	}
	if ver == "dev" {
		eztools.Debugging = true
		flagV := flag.Int("v", 1, "verbose level")
		flag.Parse()
		eztools.Verbose = *flagV
		eztools.ShowStrln("verbose " + strconv.Itoa(eztools.Verbose))
	} else {
		flagH := flag.Bool("h", false, "help messages")
		flag.Parse()
		if *flagH {
			eztools.ShowStrln("V1.0 initial release")
			eztools.ShowStrln("V1.1 account detection, verification and encryption on saving")
			eztools.ShowStrln("V2.0 import and export between files and database")
			eztools.ShowStrln("V2.1 import and export correction. Export only for manager.")
			eztools.ShowStrln("V2.2 import and export bundled to avoid writing collision.")
			return
		}
	}
	db, err = eztools.Connect()
	if err != nil {
		eztools.LogErrFatal(err)
	}
	return
}

func mainLoop(fChkMng func(db *sql.DB, acc int) (su, mng bool)) {
	logger, db := prepareEnv()
	if logger != nil {
		defer logger.Close()
	}
	if db != nil {
		defer db.Close()
	} else {
		return
	}

	upch := make(chan bool)
	svch := make(chan string)
	go eztools.AppUpgrade(db, "WeeklyPpt", ver, &svch, upch)

	choices := []string{"quit", //0
		"Fill/Edit this week's report (default)", //1
		"List this week's report",                //2
		"Generate report"}                        //3
	eztools.ShowStrln("checking for server and syncing between servers...")
	serverGot := <-upch
	if serverGot {
		<-svch
	}

	acc, fixed, err := getUsr(db)
	eztools.ShowStrln("")
	if err != nil {
		acc = eztools.InvalidID
		fixed = false
	}
	if !fixed {
		choices = append(choices, "Change account")
	}

	chkManagerNeeded := true
EXIT:
	for {
		if chkManagerNeeded {
			su, mng := fChkMng(db, acc)
			if su {
				choices = append(choices, "Maintain report structure")                 //4
				choices = append(choices, "Sync from file to database")                //5
				choices = append(choices, "Sync from file to database and write back") //6
			} else {
				if mng {
					exportPpt(db)
					break
				}
				choices = choices[:4]
			}
		}
		chkManagerNeeded = false
		c := eztools.ChooseStrings(choices)
		switch c {
		case 1, eztools.InvalidID, 2:
			if acc != eztools.InvalidID {
				break
			}
			//acc is mandatory
			acc, err = chgUsr(db)
			if err == nil || err == errNotSaved {
				chkManagerNeeded = true
			} else {
				c = 0
			}
		}
		switch c {
		case 0:
			break EXIT
		case 1, eztools.InvalidID:
			fll(db, acc)
		case 2:
			chk(db, acc)
		case 3:
			exportPpt(db)
		case 4:
			if !fixed {
				acc, err = chgUsr(db)
				if err == nil {
					chkManagerNeeded = true
				} else {
					eztools.LogErrPrint(err)
				}
			} else {
				mnt(db)
			}
		case 5:
			acc, _ = importPpt(db, acc, false)
		case 6:
			var fp *os.File
			acc, fp = importPpt(db, acc, true)
			if fp != nil {
				_, err = fp.Seek(0, io.SeekStart)
				if err == nil {
					err = fp.Truncate(0)
					if err == nil {
						wrByWR(db, fp)
					}
				}
				fp.Close()
				if err != nil {
					eztools.LogErrFatal(err)
				}
			}
		default:
			eztools.LogPrint("impossible choice: " + strconv.Itoa(c))
		}
		horizontal(0)
	}

	if serverGot {
		eztools.ShowStrln("waiting for update check to end...")
		<-upch
	}
}

func main() {
	mainLoop(chkManager)
}
